### PR TITLE
feat: testify test suite support

### LINKED ISSUE #25: Support running single test in a testify suite
When running the nearest test in a suite, I would expect the run command to only run the specified test, but instead it **quietly** misses completely.

Expected run spec command -run option
```
go test ./path/... -run '^TestSuiteName/TestName$'
```

Actual run spec command -run option
```
go test ./path/... -run '^TestName$'
```

The additional surprise here is that when then actual command is run, there is no notification that no tests were matched, the summary suggested it worked, but the coverage wasn't updated.

This is when running the nearest test, looking at the run spec produced by `build_single_test_runspec`

### PR DESCRIPTION
**Pull request by fredrikaverpil**
image

image

Please give me a hand and try it out:

```lua
  {
    "nvim-neotest/neotest",
    ft = { "go" },
    dependencies = {
      {
        "fredrikaverpil/neotest-golang",
        branch = "feat/test-suite-support",  -- ADD THIS
      },
    },
    opts = function(_, opts)
      opts.adapters = opts.adapters or {}
      opts.adapters["neotest-golang"] = {
        dap_go_enabled = true,
        testify = true,  -- ADD THIS
      }
    end,
  },


```

Aims to fix #25.

### To do
- [x] Detect receiver method name using AST/tree-sitter.
- [x] Detect suite name using AST/tree-sitter.
- [x] Create lookup between AST-detected suite name vs receiver method name.
- [x] Add hacks all over the place, so to verify my overall, initial, idea. 
- [ ] ~~Augument neotest position tree with lookup resolve for each position, prior to building runspec.~~
- [x] Generate lookup based on entire go project? (likely, yes)
- [x] Replace `pos.id`s before returning tree from `ast.lua`. Should be able to remove all added hacks.
- [x] DAP supported out of the box, piggybacking on existing code.
- [x] Verify proper output generated after test ran when running a suite.
- [ ] Trigger lookup rebuild, but based on what... 🤔
- [x] Community feedback.
  - [x] Added alternative syntax support in testify suite function.
  - [ ] ~~Table tests support.~~ Will likely not be part of initial implementation, just to realistically scope this down.
  - [x] Fixed bug on confusing regular test functions with testify suite functions.
- [x] Cleanup refactoring
  - [x] Refactor `testify.lua` into folder of files, to make it easier to become familiar with the structure/flow.
  - [x] Add types to all functions and relevant dataclasses/tables.
  - [x] Add comments/docstrings to each significant function.
  - [x] Scrutinize treesitter query/lookup generation which mixes struct/receiver nomenclature, could be confusing.
  - [ ] Consider using public Neotest methods instead of accessing private members directly (not sure if any such public methods exists).
- [x] Tests. Likely want to have a `tests/go/testify` folder/package with individual test files that tests each scenario.
  - [x] Expected lookup based on test file parsed.
  - [x] Simple test using the official testify example from their docs.
  - [x] One suite struct, several test files with test receiver methods.
  - [x] Alternative suite function syntax.
  - [x] Regular, trivial, test function should not be confused with suite function.
- [x] Settle on the naming of the option to opt-in to this behavior. Perhaps `testify_enabled`?
- [x] Write a separate docs section on this feature, as it comes with nuances, caveats and complex level of maintainability over time.

### CODE REVIEW

**Comment by nikosbatsaras** [comment]
The detection of individual tests of a Suite works well, I can see it in the Test Summary.

Notes:

- Table Test cases are not detected yet (not shown in the summary):
Screenshot 2024-07-08 at 11 54 47 AM

- Using the "Test Nearest" does not seem to work yet (not sure if this is TODO work or not). I get this symbol instead of ✅ or a ❌ :
Screenshot 2024-07-08 at 11 52 16 AM

**Comment by fredrikaverpil** [comment]
Hey thanks for giving this a spin @nikosbatsaras !

Aha, yes. Table tests. I might leave that for a little later but it should be trivial to add.

However for running the nearest tests successfully should work. You are seeing the "skipped" symbol which indicates that the test was either not executed or the adapter failed to pick up on the results in the output.

Would it be possible for you to provide a reproducible test file for me to look into?

**Comment by nikosbatsaras** [comment]
Here is a dummy test file:
```
import (
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/suite"
)

type InputTestSuite struct {
	suite.Suite
}

func TestInputTestSuite(t *testing.T) {
	s := &InputTestSuite{}
	suite.Run(t, s)
}

func (s *InputTestSuite) TestInput() {
	t := s.T()

	cases := []struct {
		Descrption string
		BodyFunc   func(map[string]any) map[string]any
	}{
		{
			"With empty rp_id",
			func(body map[string]any) map[string]any {
				body["rp_id"] = ""
				return body
			},
		},
		{
			"With empty rp_origin",
			func(body map[string]any) map[string]any {
				body["rp_origin"] = ""
				return body
			},
		},
	}

	for _, c := range cases {
		validBody := map[string]any{
			"rp_id":     "example.com",
			"rp_origin": "https://your.login.domain.example.com",
		}

		invalidBody := c.BodyFunc(validBody)

		assert.Equal(t, validBody["rp_id"], invalidBody["rp_id"])
		assert.Equal(t, validBody["rp_origin"], invalidBody["rp_origin"])
	}
}
```
and here is my `neotest-golang.lua` Lazy config file:
```
return {
    "nvim-neotest/neotest",
    ft = { "go" },
    dependencies = {
      {
        "fredrikaverpil/neotest-golang",
        branch = "feat/test-suite-support",  -- ADD THIS
      },
    },
    opts = function(_, opts)
      opts.adapters = opts.adapters or {}
      opts.adapters["neotest-golang"] = {
        go_test_args = {
          "-v",
          "-race",
          "-coverprofile=" .. vim.fn.getcwd() .. "/coverage.out",
        },
        dap_go_enabled = true,
        testify = true,  -- ADD THIS
      }
    end,
    config = function(_, opts)
      if opts.adapters then
        local adapters = {}
        for name, config in pairs(opts.adapters or {}) do
          if type(name) == "number" then
            if type(config) == "string" then
              config = require(config)
            end
            adapters[#adapters + 1] = config
          elseif config ~= false then
            local adapter = require(name)
            if type(config) == "table" and not vim.tbl_isempty(config) then
              local meta = getmetatable(adapter)
              if adapter.setup then
                adapter.setup(config)
              elseif adapter.adapter then
                adapter.adapter(config)
                adapter = adapter.adapter
              elseif meta and meta.__call then
                adapter(config)
              else
                error("Adapter " .. name .. " does not support setup")
              end
            end
            adapters[#adapters + 1] = adapter
          end
        end
        opts.adapters = adapters
      end

      require("neotest").setup(opts)
    end,
    keys = {
      { "<leader>ta", function() require("neotest").run.attach() end, desc = "[t]est [a]ttach" },
      { "<leader>tf", function() require("neotest").run.run(vim.fn.expand("%")) end, desc = "[t]est run [f]ile" },
      { "<leader>tA", function() require("neotest").run.run(vim.uv.cwd()) end, desc = "[t]est [A]ll files" },
      { "<leader>tS", function() require("neotest").run.run({ suite = true }) end, desc = "[t]est [S]uite" },
      { "<leader>tn", function() require("neotest").run.run() end, desc = "[t]est [n]earest" },
      { "<leader>tl", function() require("neotest").run.run_last() end, desc = "[t]est [l]ast" },
      { "<leader>ts", function() require("neotest").summary.toggle() end, desc = "[t]est [s]ummary" },
      { "<leader>to", function() require("neotest").output.open({ enter = true, auto_close = true }) end, desc = "[t]est [o]utput" },
      { "<leader>tO", function() require("neotest").output_panel.toggle() end, desc = "[t]est [O]utput panel" },
      { "<leader>tt", function() require("neotest").run.stop() end, desc = "[t]est [t]erminate" },
      { "<leader>td", function() require("neotest").run.run({ suite = false, strategy = "dap" }) end, desc = "Debug nearest test" },
    },
}
```

Not sure if I missed anything in the configuration 🤔

**Comment by fredrikaverpil** [comment]
Thanks for this. I'm on mobile but I see the problem. It's this part:

```go
func TestInputTestSuite(t *testing.T) {
	s := &InputTestSuite{}
	suite.Run(t, s)
}
```

The way I get all of this testify shenanigans to work at all is to somehow (and reliably) map the struct name (`InputTestSuite`) to the test function name (`TestInputTestSuite`), which is done by having a treesitter query parse the test files and produce an internal lookup. This lookup is then used to replace the receiver method with your actual suite name in the Neotest/treesitter tree.

You are using a different syntax than what I expected and I'm afraid there might be a myriad of ways to write the above snippet.

My adapter assumes you write it like this:

```go
func TestInputTestSuite(t *testing.T) {
	suite.Run(t, new(InputTestSuite)
}
```

I have no experience using testify so I'm not sure how many scenarios we are anticipating here. But key is to have one or more treesitter queries which can detect the struct name within the test function and add this data to the internal lookup that is being used.

If this feels like a lost cause we could also include a query which looks for a comment like `// testify: InputTestSuite` before or inside the `TestInputTestSuite` function, but would also be really ugly and weird to have to explain in code review to non-Neovim users. 

I'm actually not sure if treesitter is capable of inferring that `s` refers to `InputTestSuite `. This might require more sophisticated methods.

I'll play around a bit later with expanding the existing treesitter query to see what can be done.

**Comment by nikosbatsaras** [comment]
I confirm that running the Suite like so:
```
func TestPostWebauthnEntityTestSuite(t *testing.T) {
	// s := &PostWebauthnEntityTestSuite{}
	suite.Run(t, new(PostWebauthnEntityTestSuite))
}
```
makes the `Test Nearest` function work 🚀 

To sum up, 2 possible enhancements are:
1. Detect different init of Test Suites at suite.Run() (issue above ☝️)
2. Detect individual Table Test cases

I do think both are *minor* improvements (although nice to have) and should not block a possible merge, in case nothing more serious comes up.

Nice work again 👍

File: lua/neotest-golang/testify.lua [hunk 1]
```diff
@@ -0,0 +1,363 @@
...
+      (expression_statement 
+        (call_expression 
+          function: (selector_expression 
+            operand: (identifier) 
+            field: (field_identifier)) 
+          arguments: (argument_list 
+            (identifier) 
+            (identifier)))))) 
+
+  ]]
```
Comment (fredrikaverpil) [inline lua/neotest-golang/testify.lua:62]: @nikosbatsaras I think this will do. Should now support your syntax:

```go
func TestSuite(t *testing.T) {  // 
  s := &testSuiteStruct{}  // @suite_struct
  suite.Run(t, s)
}
```

Comment (nikosbatsaras) [reply lua/neotest-golang/testify.lua]: Yeap, this works 👍

Thank you ❤️

**Comment by fredrikaverpil** [comment]
@rcarriga if you are up for it, I would love your input on this PR from your perspective as the creator of Neotest. 😄 

TL;DR - I am detecting namespaces, and in the context of Go testing these will be used as [testify](https://github.com/stretchr/testify) suites. Each suite can contain any number of Go tests. But I have to do some real shenanigans to pull it all off like modifying the Neotest tree.

There have been previous attempts to support testify suites in neotest-go (not this adapter 😅) from what I understand, but with significant caveats:
* Issue, about 2 years old: https://github.com/nvim-neotest/neotest-go/issues/6
* PR, about 6 months old: https://github.com/nvim-neotest/neotest-go/pull/67

My approach looks like this:
1. Using custom treesitter query, generate a lookup which correlates testify suite structs with Go tests that should run under a testify suite.
1. Detect tests along with the Go receiver method struct as namespace (using `lib.treesitter.parse_positions`).
2. Modify the Neotest tree, so to replace the name of the namespace into the proper suite name and merge duplicate namespaces.

I am especially interested in your input of what you think about how I access private Neotest members and modify the Neotest tree in `tree_modification.lua`. Maybe there are better and more robust ways to achieve this?

**Comment by rcarriga** [comment]
Hey just catching up on this now, very nice! 
> I am especially interested in your input of what you think about how I access private Neotest members and modify the Neotest tree in tree_modification.lua. Maybe there are better and more robust ways to achieve this?

You could use lists directly instead of the tree. You can use `tree:to_list()` to convert to lists, make any changes needed, and then use `Tree.from_list` to parse it back again. I'm not sure if this is better than the current approach but at least you don't have to worry about maintaining relations such as the parent/children.

> I'd also like to know if there's any way to create callbacks to certain Neotest events, so that the lookup map can be regenerated based on e.g. the user saving a *_test.go file.

Not currently though your adapter should be called anytime a user edits a test file to parse the file again, is that not sufficient?

**Comment by fredrikaverpil** [comment]
> You could use lists directly instead of the tree. You can use tree:to_list() to convert to lists, make any changes needed, and then use Tree.from_list to parse it back again. I'm not sure if this is better than the current approach but at least you don't have to worry about maintaining relations such as the parent/children.

Sounds at least like one gain, so I'll make sure to try that out!

> Not currently though your adapter should be called anytime a user edits a test file to parse the file again, is that not sufficient?

I've added an initial trigger to generate the lookup [here](https://github.com/fredrikaverpil/neotest-golang/blob/main/lua/neotest-golang/init.lua#L19). Where do you suggest I add this `testify.lookup.generate()` in, so to re-generate the lookup whenever the user edits a test file?

**Comment by rcarriga** [comment]
> I've added an initial trigger to generate the lookup [here](https://github.com/fredrikaverpil/neotest-golang/blob/main/lua/neotest-golang/init.lua#L19). Where do you suggest I add this testify.lookup.generate() in, so to re-generate the lookup whenever the user edits a test file?

Ah OK so then if you just run that whenever `discover_positions` is called because that'll be called anytime a file is edited. One problem is however it's called per file so a large suite will trigger it many times on first startup. I would recommend adding some debouncing logic to it

**Comment by fredrikaverpil** [comment]
> One problem is however it's called per file so a large suite will trigger it many times on first startup.

Yes, exactly, this is why I haven't implemented this re-generation of the lookup yet. If I just add it into `discover_positions`, it'll get triggered way too many times in a row. If your suggestion is to implement it there and look into debouncing, I'll do that 👍  but do you think it will be robust given that the user might have `discovery.enabled = false`...? 🤔

**Comment by rcarriga** [comment]
I think debouncing is perfectly fine for this, I implemented it for the summary window to render as it has the same issue of lots of calls being made to render. It's just a simple `while true` loop using nio event objects https://github.com/nvim-neotest/neotest/blob/32ff2ac21135a372a42b38ae131e531e64833bd3/lua/neotest/consumers/summary/summary.lua#L84-L86
